pkg/controller/registry/resolver/cache/predicates.go

```diff
@@ -282,6 +282,24 @@ func (p orPredicate) String() string {
 	return b.String()
 }
 
+type notPredicate struct {
```

Comment:
I understand this section of code is added in order to write those test cases below. However, I'm pretty sure similar code is introduced in Eric's compound constraint PR. Make sure they are identical so there won't be any conflicting code later.

Reply:
Yeah, this can be thought of as provisional. If it adds value, we can keep it, if it will come from Eric, or don't need, punt it away